<a href="https://colab.research.google.com/github/yaqianhe/GIA/blob/main/Lecture02_Expressions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Lecture 2: Expressions 
## Yaqian He
## Department of Geography
## University of Central Arkansas

## 1. Constants
- **Fixed values** such as numbers, letters, and strings, are called **"constants"** because their value does not change
- Numeric **constants** are as you expect
- String **constants** use single quotes (') or double quotes (")

In [ ]:
print(123,98.6,10000,0.001)
print(123)

123 98.6 10000 0.001
123


In [ ]:
print('H',"hello world",'hello word')

H hello world hello word


## 2. Variables
- A **variable** is a named place in the memory where a programmer can store data and later retrieve the data using the **variable** "name"
- Programmers get to choose the names of the **variables**
- You can change the contents of a **variable** in a later statement

In [ ]:
x = 12.3
y = 14

In [ ]:
print(x,y)

12.3 14


In [ ]:
x = 1000

In [ ]:
print(x)

1000


## 2.1 Python variable name rules
- Must start with a letter or underscore '_'
- Must consist of letters, numbers, and underscores
- Case sensitive

**Good:** spam eggs spam23 _speed

**Bad:** 23spam #sign var.12

**Different:** spam Spam SPAM

In [ ]:
23spam = 12

SyntaxError: ignored

In [ ]:
Spam = 12
print(spam)

NameError: ignored

## 2.2 Mnemonic variable names
- Since we programmers are given a choice in how we choose our variable names, there is a bit of "best practice"
- We name variables to help us remember what we intend to store in them ("mnemonic" = "memory aid")


In [ ]:
# Example 1
xfcd1234 = 35.0
cdfgihnm = 12.50
xlq3456 = xfcd1234 * cdfgihnm
print(xlq3456)

437.5


In [ ]:
# Example 2
a = 35.0
b = 12.50
c = a * b
print(c)

437.5


In [ ]:
# Example 3
hours = 35.0
rate = 12.50
pay = hours * rate
print(pay)

437.5


## 2.3 Reserved words
You cannot use **reserved words** as variable names/identifiers

- **Reserved words examples:**

False class return is finally

None if for lambda continue

True def from while nonlocal

and del global not with

as elif try or yield

assert else import pass

break except in raise



In [ ]:
None = 12

## 3. Sentences or lines

x = 2 -------- Assignment statement

x = x + 2 -------- Assignment with expression

print(x) --------- Print statement

- **Variable:** x
- **Operator:** =, +
- **Constant:** 2
- **Function:** print()

## 3.1 Assignment statements
- We assign a value to a variable using the assignment statement (=)
- An assignment statement consists of an **expression on the right-hand side** and a **variable** to store the result.

x = 0.6 ---- A variable is a memory location used to store a value (0.6)

x = 3.9 * x * (1-x) ----- The right side is an expression. Once the expression is evaluated, the result is placed in (assigned to) x.

In [ ]:
# Example 1
x = 0.6 
x = 3.9 * x * (1-x) 
print(x)

0.9359999999999999


You could notice, the value stored in a variable (x) is updated by replacing the old value (0.6) with a new value (0.936).

## 4. Expression
- Because of the lack of mathematical symbols on computer keyboards - we use "computer-speak" to express the classic math operations
- Asterisk is multiplication
- Exponentiation (raise to a power) looks different than in math

**+** --- Addition

**-** --- Subtraction

**$*$** --- Multiplication

**/** --- Division

**$**$** --- Power

**%** --- Remainder

In [ ]:
# Example 1
x = 2
x = x + 2
print(x)

4


In [ ]:
# Example 2
y = 440 * 12
print(y)

5280


In [ ]:
# Example 3
z = y / 1000
print(z)

5.28


In [ ]:
# Example 4
j = 23
k = j % 5
print(k)

3


##4.1 Order of Evaluation
- When we string operators together - Python must know which one to do first
- This is called **"operator precedence"**
- Which operator "takes precedence" over the others?

Highest precedence rule to lowest precedence rule:
- Parentheses are always respected
- Exponentiation (raise to a power)
- Multiplication, division, and remainder
- Addition and subtraction
- Left to right

In [ ]:
# Example 1
x = 1 + (2 ** 3) / 4 * 5
print(x)

11.0


Best practice:
- Remember the rules
- When writing code - use parentheses
- When writing code - keep mathematical expressions simple enough that they are easy to understand
- Break long series of mathematical operations up to make them more clear

## 5.Type
- In Python variables, literals, and constants have a **"type"**
- Python knows the **difference** between an integer number and a string
- For example **"+"** means "addition" if something is a number and "concatenate" if something is a string

In [ ]:
# Example
stringname = "hello " + "there"
print(stringname)

hello there


## 5.1 Type matters
- Python knows what "type" everything is 
- Some operations are prohibited
- **You cannot "add 1" to a string**
- We can ask Python what type something is by using the **type()** function


In [ ]:
# Example 1
stringnumber = 'hello' + 1
print(stringnumber)

TypeError: ignored

In [ ]:
# Example 2
stringnumber = 'hello' + '1'
print(stringnumber)

hello1


In [ ]:
# Example 3
type(stringnumber)

str

## 5.2 Several types of numbers
Numbers have two main types:
- **Integers** are whole numbers: -14, -2, 0, 1, 100
- **Floating Point Numbers** have decimal parts: -2.5, 0.0, 98.6, 14.0
- There are other number types - they are variations on float and integer 

In [ ]:
# Example 1
x = 0.04
type(x)

float

In [ ]:
# Example 2
x = 4
type(x)

int

## 5.3 Numeric type conversions
- When you put an integer and floating point in an expression, the integer is implicitly converted to a float
- You can control this with the built-in functions int() and float()

In [ ]:
# Example 1
a = 42 + 54.3
print(a)
type(a)

96.3


float

In [ ]:
# Example 2
a = float(42)
print(a)
type(a)

42.0


float

In [ ]:
# Example 2
a = int(45.6)
print(a)
type(a)

45


int

For integer division produces a floating point result. Note this was different in Python 2.x


In [ ]:
# Example 1
a = 10 / 2
print(a)
type(a)

5.0


float

##5.4 String conversions
- You can also use **int()** and **float** to convert between strings and integers
- You will get an **error** if the string does not contain numeric characters

In [ ]:
# Example 1
a = '123'
b = int(a)
print(b)
type(b)

123


int

In [ ]:
# Example 1
a = '123' + 'abc'
b = int(a)
print(b)
type(b)

ValueError: ignored

## Best practice
Do comment
- Describe what is going to happen in a sequence of code
- Document who wrote the code or other ancillary information
- Turn off a line of code - perhaps temporarily

Reference: http://www.dr-chuck.com/